src/services.rs

```diff
@@ -0,0 +1,757 @@
+    jetstream,
+    jetstream::{
+        consumer::{pull::Config, Consumer},
+        stream::{Config as StreamConfig, RetentionPolicy, Source, StorageType, SubjectTransform},
+        AckKind,
+    },
+    Client,
+};
+use cloudevents::{AttributesReader, Event as CloudEvent};
+use futures::StreamExt;
+use k8s_openapi::api::core::v1::{Pod, Service, ServicePort, ServiceSpec};
+use k8s_openapi::api::discovery::v1::{Endpoint, EndpointPort, EndpointSlice};
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::OwnerReference;
+use kube::{
+    api::{Api, DeleteParams, ListParams, Patch, PatchParams},
+    client::Client as KubeClient,
+    Resource,
+};
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::net::SocketAddr;
+use std::sync::Arc;
+use tokio::sync::{mpsc, RwLock};
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, error, warn};
+use wadm::{
+    events::{Event, ManifestPublished, ManifestUnpublished},
+    model::{Manifest, TraitProperty},
+    server::{GetResult, ModelSummary},
+};
+use wash_lib::app;
+use wasmcloud_operator_types::v1alpha1::WasmCloudHostConfig;
+
+const CONSUMER_PREFIX: &str = "wasmcloud_operator_service";
+// This should probably be exposed by wadm somewhere
+const WADM_EVT_SUBJECT: &str = "wadm.evt";
+const OPERATOR_STREAM_NAME: &str = "wasmcloud_operator_events";
+
+/// Commands that can be sent to the watcher to trigger an update or removal of a service.
+#[derive(Clone, Debug)]
+enum WatcherCommand {
+    UpsertService(ServiceParams),
+    RemoveService {
+        name: String,
+        namespaces: HashSet<String>,
+    },
+    RemoveServices {
+        namespaces: HashSet<String>,
+    },
+}
+
+/// Parameters for creating or updating a service in the cluster.
+#[derive(Clone, Debug)]
+pub struct ServiceParams {
+    name: String,
+    namespaces: HashSet<String>,
+    lattice_id: String,
+    port: u16,
+    host_labels: Option<HashMap<String, String>>,
+}
+
+/// Watches for new services to be created in the cluster for a partcular lattice and creates or
+/// updates them as necessary.
+#[derive(Clone, Debug)]
+pub struct Watcher {
+    namespaces: HashSet<String>,
+    lattice_id: String,
+    nats_client: Client,
+    shutdown: CancellationToken,
+    consumer: Consumer<Config>,
+    tx: mpsc::Sender<WatcherCommand>,
+}
+
+impl Drop for Watcher {
+    fn drop(&mut self) {
+        self.shutdown.cancel();
+    }
+}
+
+impl Watcher {
+    /// Creates a new watcher for a particular lattice.
+    fn new(
+        namespace: String,
+        lattice_id: String,
+        nats_client: Client,
+        consumer: Consumer<Config>,
+        tx: mpsc::Sender<WatcherCommand>,
+    ) -> Self {
+        let watcher = Self {
+            namespaces: HashSet::from([namespace]),
+            nats_client,
+            lattice_id: lattice_id.clone(),
+            consumer,
+            shutdown: CancellationToken::new(),
+            tx,
+        };
+
+        // TODO is there a better way to handle this?
```

Comment:
That won't work without a pretty serious refactor, since the `ServiceWatcher` relies on some methods and fields on the `Watcher` struct that would need to either be duplicated across both the `Watcher` and `WatcherHandles`. Might just be a symptom of bad design, but cloning the underlying `Watcher` seems like the least bad option at the moment